### ISSUE TITLE
[Java - Jersey2] Issue with deserialize when response is null

### ISSUE THREAD

**Issue by ukeshava** [swagger-api/swagger-codegen#6226]
Generated ApiClient has deserialize function as below:
```
public <T> T deserialize(Response response, GenericType<T> returnType) throws ApiException {
    if (response == null || returnType == null) {
      return null;
    }

    if ("byte[]".equals(returnType.toString())) {
      // Handle binary response (byte array).
      return (T) response.readEntity(byte[].class);
    } else if (returnType.getRawType() == File.class) {
      // Handle file downloading.
      T file = (T) downloadFileFromResponse(response);
      return file;
    }

    String contentType = null;
    List<Object> contentTypes = response.getHeaders().get("Content-Type");
    if (contentTypes != null && !contentTypes.isEmpty())
      contentType = String.valueOf(contentTypes.get(0));
    if (contentType == null)
      throw new ApiException(500, "missing Content-Type in response");

    return response.readEntity(returnType);
  }
```

Since content-type is missing in response (response is null), hitting the exception:
"missing Content-Type in response"

However, if I remove the below code:
 if (contentType == null)
      throw new ApiException(500, "missing Content-Type in response");

it runs successfully and returns null. 

Can we remove the below lines of code from being generated:
```
String contentType = null;
    List<Object> contentTypes = response.getHeaders().get("Content-Type");
    if (contentTypes != null && !contentTypes.isEmpty())
      contentType = String.valueOf(contentTypes.get(0));
    if (contentType == null)
      throw new ApiException(500, "missing Content-Type in response");
```

Reason:
1. It is not necessary for handling null as readEntity still returns null without exception in case of response being null
2. For valid content-type, deserialize funciton anyways considers return type for readEntity operation and not the content-type

Swagger-codegen version: 2.2.2
gradle code block used to generate code:
```
task generateSwaggerClient(dependsOn: ':swagger') { //task that generates swagger.json
    inputs.file("$swaggerSourceFile")
    outputs.dir("$projectDir/$swaggerTargetFolder")
    doLast {
        def config = new CodegenConfigurator()
        config.setInputSpec("$swaggerSourceFile")
        config.setOutputDir("$projectDir")
        config.setLang('java')
        config.setAdditionalProperties([
        'invokerPackage'   : 'io.swagger.client.jersey2',
        'modelPackage'     : 'io.swagger.client.jersey2.model',
        'apiPackage'       : 'io.swagger.client.jersey2.api',
        'library'          : 'jersey2',
        'serializableModel': 'true',
        'dateLibrary'      : 'java8',
        'interfaceOnly' : 'true',
        'sourceFolder'  : swaggerTargetFolder
        ])
    new DefaultGenerator().opts(config.toClientOptInput()).generate()
    }
}
```

**Comment by wing328** [comment]
@ukeshava thanks for reporting the issue. I wonder if you can also try the `okhttp-gson` or `retrofit2` Java API client to see if the same issue occurs.

**Comment by ukeshava** [comment]
With okhttp-gson, it takes default content-type (application/json) when content-type is null.

```
String contentType = response.headers().get("Content-Type");
        if (contentType == null) {
            // ensuring a default content type
            contentType = "application/json";
        }
        if (isJsonMime(contentType)) {
            return json.deserialize(respBody, returnType);
        } else if (returnType.equals(String.class)) {
            // Expecting string, return the raw response body.
            return (T) respBody;
        } else {
            throw new ApiException(
                    "Content type \"" + contentType + "\" is not supported for type: " + returnType,
                    response.code(),
                    response.headers().toMultimap(),
                    respBody);
        }
```

But getting a different error while compiling generated code with okhttp-gson:
   src\generated\java\io\swagger\client\auth\OAuthOkHttpClient.java:58: error: no suitable method found for createCustomResponse(String,String,int,Class<T#1>)
            return OAuthClientResponseFactory.createCustomResponse(
                                             ^
    method OAuthClientResponseFactory.<T#2>createCustomResponse(String,String,int,Map<String,List<String>>,Class<T#2>) is not applicable
      (cannot infer type-variable(s) T#2
        (actual and formal argument lists differ in length))
    method OAuthClientResponseFactory.<T#3>createCustomResponse(InputStream,String,int,Map<String,List<String>>,Class<T#3>) is not applicable
      (cannot infer type-variable(s) T#3
        (actual and formal argument lists differ in length))
  where T#1,T#2,T#3 are type-variables:
    T#1 extends OAuthClientResponse declared in method <T#1>execute(OAuthClientRequest,Map<String,String>,String,Class<T#1>)
    T#2 extends OAuthClientResponse declared in method <T#2>createCustomResponse(String,String,int,Map<String,List<String>>,Class<T#2>)
    T#3 extends OAuthClientResponse declared in method <T#3>createCustomResponse(InputStream,String,int,Map<String,List<String>>,Class<T#3>)
Note: C:\hydra\swaggerClient\src\generated\java\io\swagger\client\jersey1\ApiClient.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
1 error

Would like to be able to use jersey2, as we are using jersey client for other testing as well. Also, have distributed swaggerClient for other teams. Could you please help?

**Comment by wing328** [comment]
@ukeshava please submit a PR with the following removed:
```
    if (contentType == null)
      throw new ApiException(500, "missing Content-Type in response");
```
and then the CIs can verify the change as well.

**Comment by andreeahash** [comment]
Any updates on this issue? 
We have problems with the code generated for APIs that return http code 201 with no content.

**Comment by wing328** [comment]
@andreeahash Can you try the suggested fix mentioned above (https://github.com/swagger-api/swagger-codegen/issues/6226#issuecomment-319910992) and submit a PR accordingly?

**Comment by Usnul** [comment]
#6977 
I do not have the environment set up locally. So i am unable to run the suggested checks.

**Comment by wing328** [comment]
@Usnul we've released stable version 2.3.0, which includes the fix: https://twitter.com/wing328/status/944259218386718720

Please check it out and let us know if you still experience issue.
